src/commands/dataSourceCommand.ts

```diff
@@ -214,52 +214,61 @@ export async function populateScratchpad(
 export async function runDataSource(
   dataSourceForm: DataSourceFiles
 ): Promise<void> {
-  Object.assign(ext.insightsMeta, await getMeta());
-  if (!ext.insightsMeta.assembly) {
-    ext.outputChannel.appendLine(
-      `To run a datasource you need to be connected to an Insights server`
-    );
-    window.showErrorMessage(
-      "To run a datasource you need to be connected to an Insights server"
-    );
+  if (DataSourcesPanel.running) {
     return;
   }
+  DataSourcesPanel.running = true;
 
-  dataSourceForm.insightsNode = getConnectedInsightsNode();
-  const fileContent = dataSourceForm;
+  try {
+    Object.assign(ext.insightsMeta, await getMeta());
+    if (!ext.insightsMeta.assembly) {
+      ext.outputChannel.appendLine(
+        `To run a datasource you need to be connected to an Insights server`
+      );
+      window.showErrorMessage(
+        "To run a datasource you need to be connected to an Insights server"
+      );
+      return;
+    }
 
-  let res: any;
-  const selectedType = getSelectedType(fileContent);
-  ext.isDatasourceExecution = true;
-  switch (selectedType) {
-    case "API":
-      res = await runApiDataSource(fileContent);
-      break;
-    case "QSQL":
-      res = await runQsqlDataSource(fileContent);
-      break;
-    case "SQL":
-    default:
-      res = await runSqlDataSource(fileContent);
-      break;
-  }
+    dataSourceForm.insightsNode = getConnectedInsightsNode();
+    const fileContent = dataSourceForm;
+
+    let res: any;
+    const selectedType = getSelectedType(fileContent);
+    ext.isDatasourceExecution = true;
+    switch (selectedType) {
+      case "API":
+        res = await runApiDataSource(fileContent);
+        break;
+      case "QSQL":
+        res = await runQsqlDataSource(fileContent);
+        break;
+      case "SQL":
+      default:
+        res = await runSqlDataSource(fileContent);
+        break;
+    }
 
-  ext.isDatasourceExecution = false;
-  if (res.error) {
-    window.showErrorMessage(res.error);
-  } else if (ext.resultsViewProvider.isVisible()) {
-    writeQueryResultsToView(
-      res,
-      getQuery(fileContent, selectedType),
-      selectedType
-    );
-  } else {
-    const resString = arrayToTable(res);
-    writeQueryResultsToConsole(
-      resString,
-      getQuery(fileContent, selectedType),
-      selectedType
-    );
+    ext.isDatasourceExecution = false;
+    if (res.error) {
+      window.showErrorMessage(res.error);
+    } else if (ext.resultsViewProvider.isVisible()) {
+      writeQueryResultsToView(
+        res,
+        getQuery(fileContent, selectedType),
+        selectedType
+      );
+    } else {
+      const resString = arrayToTable(res);
+      writeQueryResultsToConsole(
+        resString,
+        getQuery(fileContent, selectedType),
+        selectedType
+      );
+    }
+  } finally {
```

Comment:
also add a catch, for in case error, do ther DataSourcePanel.running = false

Reply:
errors will be catched in upper layers, (just as before) the finally. block handles proper clean up